### PR TITLE
add bfd check from lrp to  underlay gw

### PR DESCRIPTION
**Pull request by bobz965**
# Pull Request

- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Features

add bfd check from lrp to  underlay gw

## Which issue(s) this PR fixes

Fixes #(issue-number)

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 0bbd559</samp>

This pull request adds support for BFD monitoring for the EIP SNAT feature in the kube-ovn controller and the OVN-NB client. It also adds and updates mock types, methods, and unit tests for BFD and gateway chassis objects. It also fixes some minor issues in error handling and parameter naming.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 0bbd559</samp>

> _We're the vpc crew and we sail the OVN_
> _We monitor the BFD and the gateway chassis_
> _We list and update and delete them as we please_
> _Heave away, heave away, heave away on three_

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 0bbd559</samp>

*  Add methods and logic for BFD monitoring and updating for EIP SNAT feature ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3334-R3348), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3364-R3393), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3589-R3603), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3864-R3875), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R4187-R4205), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R4250-R4268), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1L1044-R1052), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR323-R331), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL533-R552), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL550-R583), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1097-R1140), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-50fe35ba11ad04edb53bfaccef915df3650a81844495ed73d3c158e975c04c49L42-R57), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L6-R16), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L26-R63), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L35-R71), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L57-R113), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L69-R124), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7R136-R331), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L42-R43), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L60-R72), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L85-R86), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L102-R107), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L116-R116))
  * Add `ListBFD`, `ListDownBFDs`, `ListUpBFDs`, `MonitorBFD` and `UpdateBFD` methods to `NbClient` interface and `MockBFD` interface ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3334-R3348), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3364-R3393), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3589-R3603), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R3864-R3875), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R4187-R4205), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R524-R599))
  * Implement `ListBFD`, `ListDownBFDs`, `ListUpBFDs`, `MonitorBFD` and `UpdateBFD` methods in `OVNNbClient` struct and add imports and constants for BFD logic ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L6-R16), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L26-R63), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L35-R71), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L57-R113), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L69-R124), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7R136-R331))
  * Modify return type of `ListBFD` method from slice to pointer to slice and update length checks and index accesses accordingly ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L35-R71), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-63bbdb9bc1a7cbb8dea5e9d908c83a21c1781b2ced6f2ce1881de811e22b2de7L69-R124), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L60-R72), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L102-R107), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L116-R116))
  * Modify destination IP addresses used in `testListBFD` and `testDeleteBFD` functions to avoid conflict with other tests ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L42-R43), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-4853d5c459827eb1c2b6e7bee9216ecb3ad565644692abbbc753067f875c8c27L85-R86))
  * Add condition to check if EIP SNAT feature is enabled before calling `MonitorBFD` method in `startWorkers` function ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1L1044-R1052))
  * Add block to get external subnet object if EIP SNAT feature is enabled in `handleAddOrUpdateVpc` function ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR323-R331))
  * Add condition to check if vpc is not default vpc and external subnet is not logical gateway in `handleAddOrUpdateVpc` function ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL533-R552))
  * Add block to create BFD objects between logical router port and physical switch gateway if vpc enables BFD and external subnet is not logical gateway in `handleAddOrUpdateVpc` function ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL550-R583))
  * Add block to delete BFD objects for logical router port in `handleAddVpcExternalSubnet` function ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1097-R1140))
  * Add method for `ListGatewayChassisByLogicalRouterPort` to `LogicalRouterPort` interface and `MockLogicalRouterPort` interface ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-50fe35ba11ad04edb53bfaccef915df3650a81844495ed73d3c158e975c04c49L42-R57), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R357-R371))
* Add methods and logic for gateway chassis updating based on BFD status ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R430-R471), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R4250-R4268), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-50fe35ba11ad04edb53bfaccef915df3650a81844495ed73d3c158e975c04c49R199), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R14), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R31-R46), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R78-R100), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7L84-R125), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R142-R144), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7L106-R150), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7L116-R160))
  * Add mock type for `GatewayChassis` and its recorder ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R430-R471))
  * Add `GatewayChassis` interface to `NbClient` interface ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-50fe35ba11ad04edb53bfaccef915df3650a81844495ed73d3c158e975c04c49R199))
  * Add `UpdateGatewayChassis` method to `NbClient` interface and `OVNNbClient` struct ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71R4250-R4268), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R31-R46))
  * Add `ListGatewayChassisByLogicalRouterPort` method to `OVNNbClient` struct ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R78-R100))
  * Modify parameter name of `newGatewayChassis` method from `gwChassisName` to `lrpName` and add `lrpName` parameter to the method ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7L84-R125), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R142-R144), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7L116-R160))
  * Modify parameter name of `CreateGatewayChassisesOp` method from `lrpName` to `lrName` ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7L106-R150))
  * Add import for util package and use constant for maximum priority of gateway chassis ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-9303d6d8c4b4c52f1cd1cf6a663085b3e9d3515ca917667437e20feaf778e5e7R14), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-23fe1930ce750ec27bfa79a8fb8d076245c64df2b62ea6d6367cdd59260b164cR153-R154))
* Add error logging and wrapping to various methods in `ovn-nb-logical_router_port.go` file ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L53-R55), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L116-R126), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L153-R166), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L168-R178), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L183-R194), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06R204), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L198-R212), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L214-R230), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L228-R245), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L242-R261), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06R269-R273), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06L289-R315), [link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-e6855e2616a51a054e96fd9081cc8852239a14f67ec293fb75646a3c3e819f06R321))
* Add condition to check if node name is not empty and not duplicated before appending it to `gwNodes` slice in `getGatewayChassis` function ([link](https://github.com/kubeovn/kube-ovn/pull/3441/files?diff=unified&w=0#diff-945b73ad15889230d50be6f6fbd9e7352e96338574ce6f8a53b7a32b2e334739L229-R231))

### CODE REVIEW

File: pkg/controller/vpc.go [hunk 1]
```diff
@@ -317,9 +317,19 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
 		return err
 	}
 
+	var externalSubnet *kubeovnv1.Subnet
+	isExternalSubnetExist := false
```
Comment (oilbeater) [inline pkg/controller/vpc.go:321]:
```suggestion
	externalSubnetExist := false
```

File: pkg/controller/vpc.go [hunk 2]
```diff
@@ -317,9 +317,19 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
+	if c.config.EnableEipSnat && c.config.NetworkType != util.NetworkTypeVlan {
```
Comment (oilbeater) [inline pkg/controller/vpc.go:322]: `c.config.NetworkType` only influences the default subnet, the custom VPC will not influenced by this config

File: pkg/ovs/ovn-nb-bfd.go [hunk 3]
```diff
@@ -3,13 +3,18 @@ package ovs
...
 
 	"github.com/kubeovn/kube-ovn/pkg/ovsdb/ovnnb"
+	"github.com/kubeovn/kube-ovn/pkg/util"
+
+	"github.com/ovn-org/libovsdb/cache"
+	"github.com/ovn-org/libovsdb/model"
 )
 
-func (c *OVNNbClient) ListBFD(lrpName, dstIP string) ([]ovnnb.BFD, error) {
+func (c *OVNNbClient) ListBFD(lrpName, dstIP string) (*[]ovnnb.BFD, error) {
```
Comment (oilbeater) [inline pkg/ovs/ovn-nb-bfd.go:17]: why use a pointer to a slice here

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/ovs/ovn-nb-bfd.go [hunk 4]
```diff
@@ -66,19 +128,219 @@ func (c *OVNNbClient) DeleteBFD(lrpName, dstIP string) error {
...
+			klog.Errorf("failed to check bfd status for lrp %s dst ip %s, %v", bfd.LogicalPort, bfd.DstIP, err)
+			continue
+		} else if ok {
+			break
+		}
+	}
+}
+
+func (c *OVNNbClient) bfdUpdateL3HAHandler(table string, oldModel, newModel model.Model) {
+	if table != "BFD" {
```
Comment (zhangzujian) [inline pkg/ovs/ovn-nb-bfd.go:225]:
```suggestion
	if table != ovnnb.BFDTable {
```

File: pkg/ovs/ovn-nb-bfd.go [hunk 5]
```diff
@@ -66,19 +128,219 @@ func (c *OVNNbClient) DeleteBFD(lrpName, dstIP string) error {
...
+			} else if ok {
+				break
+			}
+			try++
+		}
+	}
+}
+
+func (c *OVNNbClient) bfdDelL3HAHandler(table string, model model.Model) {
+	if table != "BFD" {
```
Comment (zhangzujian) [inline pkg/ovs/ovn-nb-bfd.go:341]:
```suggestion
	if table != ovnnb.BFDTable {
```

File: pkg/ovs/ovn-nb-bfd.go [hunk 6]
```diff
@@ -66,19 +128,219 @@ func (c *OVNNbClient) DeleteBFD(lrpName, dstIP string) error {
...
+		return true, nil
+	}
+	// bfd status is still down
+	err = fmt.Errorf("lrp %s bfd dst ip %s status is down", lrpName, bfd.DstIP)
+	klog.Error(err)
+	return false, err
+}
+
+func (c *OVNNbClient) bfdAddL3HAHandler(table string, model model.Model) {
+	if table != "BFD" {
```
Comment (zhangzujian) [inline pkg/ovs/ovn-nb-bfd.go:196]:
```suggestion
	if table != ovnnb.BFDTable {
```

File: pkg/ovs/ovn-nb-bfd.go (hunk 3 above)
Comment (bobz965) [reply pkg/ovs/ovn-nb-bfd.go:17]: I used to think that a pointer always has a better performance, but not always actually. i will recover it.

File: pkg/controller/vpc.go (hunk 2 above)
Comment (bobz965) [reply pkg/controller/vpc.go:322]: the vlan e2e seems config EnableEipSnat as true when use vlan, so the e2e failed.
and also, the NetworkTypeVlan means not use vpc case , so custom vpc will not be used if the cluster use vlan default network.  if use custom vpc means the cluster not use vlan default network

File: pkg/controller/vpc.go (hunk 1 above)
Comment (bobz965) [reply pkg/controller/vpc.go]: fixed

File: pkg/controller/vpc.go (hunk 2 above)
Comment (bobz965) [reply pkg/controller/vpc.go]: fixed

File: pkg/ovs/ovn-nb-bfd.go (hunk 3 above)
Comment (bobz965) [reply pkg/ovs/ovn-nb-bfd.go]: fixed

File: pkg/ovs/ovn-nb-bfd.go (hunk 5 above)
Comment (bobz965) [reply pkg/ovs/ovn-nb-bfd.go]: fixed

**Review by oilbeater — APPROVED** [review]

**Pull request merged**